### ISSUE TITLE
What Does Accessibility Look Like in Baseline

### ISSUE THREAD

**Issue by lolaodelola** [web-platform-dx/web-features#3789]
_As per 2026 focus area 2 '[Integrate Accessibility into Feature Status](https://docs.google.com/document/d/1e58q9LsObhg2vuORHntWZsH5nOPldlAvCQ5lBnvqt-E/edit?tab=t.0#heading=h.wbn40qtfz2nc)'. & following on from [today's conversation](https://docs.google.com/document/d/1ree75ImLZjf60lTZ3BhCaLHygxgywr7SBXp-q0xPs8A/edit?tab=t.tprqk9av7dp9#heading=h.fj6ahxp3j8r8)_

- We should understand what accessibility in Baseline looks like with &/or without ACD.
- In order to understand what Baseline could look like with ACD, we need ACD data.
- We should explore creating [accessibility mappings ](https://github.com/web-platform-dx/web-features-mappings)
- Accessibility data could be provided as an add-on to Baseline, where there could potentially be room for more nuance on accessibility statuses
- [html5accessibility](https://html5accessibility.com/) is an example of giving yes/no/partial support statuses for accessibility features
- ARIA-AT has [started tracking support for HTML-AAM ](https://aria-at.w3.org/reports/html-features)features.
- We're currently using WCAG [accessibility-supported](https://www.w3.org/TR/WCAG21/#dfn-accessibility-supported) definition and paraphrasing it to: "the browser has to expose the web feature to the accessibility tree and be supported by the AT".
- There are test gaps in both WPT and ARIA-AT we'll need to fill.
- ACD team will be working more on how to compute support in the ACD Collector

**Comment by lolaodelola** [comment]
The ACD stakeholders have been discussing amongst ourselves about how we should compute support statuses, is this a discussion this group should be involved with as well? The goal is to have data that follows the same shape as BCD as closely as possible.

**Comment by japui-coder** [comment]
What Does Accessibility Look Like in Baseline #3507

**Comment by lolaodelola** [comment]
On April 2nd we discussed getting a subset of WPT tests to simulate what the data could look like/issues that could arise. Many of the WPTs have web feature files that map feature keys to BCD keys, [for example, this is the web feature metadata file for the `li` element](https://github.com/web-platform-tests/wpt/blob/db01946c632e4e2b5257c291e061f0595a29878c/html/semantics/grouping-content/the-li-element/WEB_FEATURES.yml).

Assuming we already have our tests, here are two patterns of issues I've seen so far:

1. **Web APIs which are not accessibility-supported by default but are widely available.** 
Canvas is a good example, in the canvas spec there is acknowledgement that canvas shouldn't be used without a fallback, MDN advises against using canvas for accessible websites (which should be most of the web). Accessibility data would say this feature is *not* widely available or supported because the information rendered in a canvas is hidden from the accessibility tree. So technically the API is widely available in that browsers have implemented it to the best of their ability, however, it's not accessibility-supported because the spec itself doesn't have anyway of showing canvas-rendered elements to the a-tree.

2. **Web APIs which are inconsistently implemented across browsers**
Typically we don't care about implementation details, however, how a browser chooses to implement can have repercussions on users. Popover is a good example. The `popover` attribute has built-in accessibility mechanics which means that it makes the state & contents of the popover available to the a-tree using `aria-expanded` and `aria-details` to the relevant elements and also adds a `role=group` to popover content where necessary. This happens in all major browsers except in Safari where `aria-details` isn't added automatically to the relevant element. Web developers can manually add the `aria-details` themselves so that users using Safari with other popular screen-readers can benefit, **but Safari breaks convention & expectation here**.

**Comment by lolaodelola** [comment]
Summary from the meeting:

1. We still need to discuss scenarios where ACD and Baseline disagree on what is widely available for web features which can't be accessible because they don't have accessibility semantics.
2. For web features which break accessibility norms or spec requirements, we can regress baseline status.

**Comment by ddbeck** [comment]
This came up on yesterday's WebDX call and I've been thinking it over a bit since.

As I heard it, Lola was making the case that a category 1 feature, like canvas, can have accessibility hazards _by specification_ that make it unsuitable for developers to use. In that case, I think we already have a mechanism for conveying that information to developers: discouraged features. Baseline already accounts for such features by allowing spec authors to unilaterally disqualify a feature from Baseline. If the spec editors were to add some text to the spec explicitly discouraging the use of  `<canvas>` on the web, we could represent that information in [`canvas.yml`](https://github.com/web-platform-dx/web-features/blob/main/features/canvas.yml) today.

A category 2 feature sounds like a `partial_implementation` in BCD (or at very least, a note, depending on how permissive the spec is). In the narrow case of setting HTML attributes, these are directly exposed to developers and could be represented in BCD. For popover specifically, I'd welcome an issue on BCD (or a link to one) to see how we might capture that information.

I'd expect the data that would be novel and most interesting for setting a Baseline status would be in a hidden category 3: things that work in the realm of web content (i.e., anything that you can modify, inspect, or respond to with existing authoring, DOM APIs, media queries, and the like), but for whatever reason break at the layer of the accessibility tree (which is unexposed to web content) or assistive technology (doubly unexposed to web content, e.g., a screen reader failing to do something with a correctly-formed accessibility tree). Naturally, this is the hardest thing to get data for!

**Comment by lolaodelola** [comment]
@ddbeck This makes sense to me.

> Baseline already accounts for such features by allowing spec authors to unilaterally disqualify a feature from Baseline. If the spec editors were to add some text to the spec explicitly discouraging the use of <canvas> on the web, we could represent that information in [canvas.yml](https://github.com/web-platform-dx/web-features/blob/main/features/canvas.yml) today.

Would the following meet the discouragement level for Baseline?

>When authors use the [canvas](https://html.spec.whatwg.org/#the-canvas-element) element, they must also provide content that, when presented to the user, conveys essentially the same function or purpose as the [canvas](https://html.spec.whatwg.org/#the-canvas-element)'s bitmap. This content may be placed as content of the [canvas](https://html.spec.whatwg.org/#the-canvas-element) element. The contents of the [canvas](https://html.spec.whatwg.org/#the-canvas-element) element, if any, are the element's [fallback content](https://html.spec.whatwg.org/#fallback-content).

What is the messaging if a feature goes from widely available to discouraged for web developers?

>For popover specifically, I'd welcome an issue on BCD (or a link to one) to see how we might capture that information.

I can open that issue

>I'd expect the data that would be novel and most interesting for setting a Baseline status would be in a hidden category 3: things that work in the realm of web content (i.e., anything that you can modify, inspect, or respond to with existing authoring, DOM APIs, media queries, and the like), but for whatever reason break at the layer of the accessibility tree (which is unexposed to web content) or assistive technology (doubly unexposed to web content, e.g., a screen reader failing to do something with a correctly-formed accessibility tree). 

Last year Adrian opened a series of issues which meet this criteria (I can't find them now) but essentially, Voice Over in (only) Safari was not speaking the preceding text in `<del>`, `<mark>`, `<i>`, `<s>`, this was an issue with Voice Over and the fix required an OS update for users. Ideally we'd want to let web developers know that using these HTML elements can cause missing text for a subsection of users. This is an issue that we'd hope to catch with the screen-reader side of testing with ACD.

**Comment by lolaodelola** [comment]
The BCD issue is here: https://github.com/mdn/browser-compat-data/issues/29566

**Comment by captainbrosset** [comment]
Regarding canvas, I don't feel good about discouraging the feature in Baseline. The spec text doesn't really discourage against using the feature, it says that authors must follow best practices to avoid accessibility issues.
So far, we've been using the discouraged status only for features which specs specifically mark as obsolete, or similar.

Canvas is problematic because you can use it to create inaccessible interactive UIs, but also to create artwork for illustration purposes, and have the right fallback content nested in the element. So, it's a case of a feature that's hard to make accessible, and should probably be avoided most of the times, but still has real use cases, for which no alternatives exist (and, which might, over time, become more accessible as we invent new ways to expose canvas content to the a-tree).

So, for category 1 issues, I think notes is what we'd need. A way to educate users on best practices that they must follow if they want to use the feature.

**Comment by captainbrosset** [comment]
Category 2 issues feel like they should be treated as bugs, at least when the spec allows us to do so.

popover is a special case apparently because the spec doesn't actually require implementations to add the correct aria attributes.
Because of this, I think popover actually belongs to category 1, and we could add a note telling devs what to do to ensure their popovers are accessible.

For other category 2 cases, we should make sure the partial/inconsistent implementations are recorded in BCD as partials, so that we don't give the features a Baseline status.

**Comment by captainbrosset** [comment]
Attempting to summarize what I'm getting from this thread so far:

- Category 1: web features which require accessibility guidelines for developers to create accessible experiences.
  Baseline status still applies, but should come with developer guidance. Perhaps a note. Perhaps a link to existing guidance.
  MDN's accessibility sections come to mind as the right source of truth for this. Maybe we should map to this.

- Category 2: web features that aren't accessibility supported because they don't expose the right information to the accessibility tree.
  This should be treated as a partial implementation in BCD, preventing Baseline status.

- Category 3: web features that should be accessibility supported, but aren't.
  That's the part that requires ACD data, and if we did have it, I think it should prevent features from getting the Baseline status.

Am I missing anything, or have I gotten any of this wrong?

**Comment by captainbrosset** [comment]
Here's an example for category 3 issues, I think: https://github.com/mdn/browser-compat-data/issues/29610
Here, the `ariaNotify` API (which happens to be an accessibility API, btw), doesn't work properly with Voice Over (see [specific comment](https://github.com/mdn/browser-compat-data/issues/29610#issuecomment-4383439210)).

This is a case of a feature that's not properly supported at the AT level, but should, and which requires manual testing to verify.

**Comment by ddbeck** [comment]
> Am I missing anything, or have I gotten any of this wrong?

I think you have this correct, with one asterisk.

>  * Category 1: web features which require accessibility guidelines for developers to create accessible experiences.
>  Baseline status still applies, but should come with developer guidance. Perhaps a note. Perhaps a link to existing guidance.
>  MDN's accessibility sections come to mind as the right source of truth for this. Maybe we should map to this.

I think this is correct and I like this idea a lot. It seems to me that something like "there are X accessibility notices associated with feature" would be a cool thing to surface to developers. Maybe fodder for https://github.com/web-platform-dx/web-features-mappings?

> * Category 2: web features that aren't accessibility supported because they don't expose the right information to the accessibility tree.
> This should be treated as a partial implementation in BCD, preventing Baseline status.

This part is challenging. There's a big question mark here, as I'm not sure there's an authoritative source for what constitutes the 'right' information and (even if there was) how we would know whether the tree actually represented that information faithfully.

BCD treats a web page as the observable universe, where only the things you can do with the JS, APIs, CSS, HTTP, and so on that have an observable effect on the page or browser users can be represented. As an abstraction for other non-browser software, the accessibility tree is dark matter. We can only observe the tree indirectly by the behavior of non-browser software which might have its own bugs, implementer-discretionary behavior, and so on.

(The exception here is dev tools, but dev tools have no specified compatibility expectations or a practical testing surface that would make data about it legible and maintainable, so it's not in scope for BCD.)

> * Category 3: web features that should be accessibility supported, but aren't.
> That's the part that requires ACD data, and if we did have it, I think it should prevent features from getting the Baseline status.

Yes, agreed.

**Comment by lolaodelola** [comment]
> BCD treats a web page as the observable universe, where only the things you can do with the JS, APIs, CSS, HTTP, and so on that have an observable effect on the page or browser users can be represented. As an abstraction for other non-browser software, the accessibility tree is dark matter. We can only observe the tree indirectly by the behavior of non-browser software which might have its own bugs, implementer-discretionary behavior, and so on.

We can definitely write browser tests that check the accessibility tree in WPT. In the case of Canvas, we could write tests that check to see if the content is represented in the A-Tree, but we also know that it won't be in the A-tree because it's documented in the spec for Canvas.

**Comment by ddbeck** [comment]
> We can definitely write browser tests that check the accessibility tree in WPT.

Can you link to an example for this? I'd like to see what exactly it's checking (it might have some implications for the BCD collector, depending on the details).

**Comment by lolaodelola** [comment]
Ah, I misspoke, I said:

> In the case of Canvas, we could write tests that check to see if the content is represented in the A-Tree

We can't just yet, but the work is happening to change that: https://github.com/WICG/aom/issues/203 It's a very long thread, but in a nutshell, they are actively working on expanding the webdriver mechanics to include an accessible node feature which will return a representation of the testable element.

My point still stands that in the case of Canvas we don't need to depend only on the AT output for what's expected in the browser, as there's documentation (through GH conversations and spec language) that indicates the authors/editors of the spec know the feature is inaccessible. And typically, that is the case. We could potentially consider looking at the accessibility horizontal reviews  & TAG reviews for web features too which would also highlight accessibility failures.

> Can you link to an example for this?

Sure, right now we write tests for computed role and computed label/name (as defined in HTML-AAM or other doc): https://github.com/web-platform-tests/wpt/blob/master/html-aam/roles.html
